docs/contributing/Powershell Guidelines.md

```diff
@@ -0,0 +1,99 @@
+Powershell Guidelines
+==============
+
+Powershell is primarily used in this repo as the backbone of our infrastructure.  As such 
+Powershell scripts need to be portable, reliable and take advantage of stop on error 
+approaches.  The guidelines below are meant to push scripts to this mindset of execution.
+
+# Coding Style
+
+1. Opening braces always go at the end of an expression / statement. 
+1. Closing braces always occur on an otherwise empty line.
+1. We use four space indentation.
+1. Use Pascal casing for functions and where possible follow the Verb-Name convention.
+1. Use Camel casing for all other identifier.
+1. Use full command names instead of aliass.  For example `Get-Content` vs. `gc`.  Aliases can be 
+overriden by the environment and hence are not considered portable. 
+
+# General Guidelines
+
+## Body 
+
+All scripts shall include the following two statements after parameter declarations:
+
+``` powershell
+Set-StrictMode -version 2.0
+$ErrorActionPreference="Stop"
+```
+
+This both forces Powershell into a more strict mode of interepretation and swaps out the 
+"On Error Resume Next" approach for an "On Error Stop" model.  Both of these help with our
+goals of reliability as it makes errors hard stops (usless specifically stated otherwise)
+
+The body of a Powershell script shall be wrapped inside the following try / catch template:
+
+``` powershell
+try { 
+    # Body of Powershell script goes here
+}
+catch {
+    Write-Host $_
+    Write-Host $_.Exception
+    exit 1
+}
+```
+
+This will force scripts to exit with an error code when an unhandled exception occurs. 
+
+## Parameters
+
+The parameter block shall occur at the top of the script.  Authors should consider disabling
+positional binding:
+
+``` powershell
+[CmdletBinding(PositionalBinding=$false)]
+param (
+    [switch]$test64 = $false,
+    [switch]$testDeterminism = $false)
+```
+
+This helps alert callers to casual typos, which would otherwise go unnoticed, by making it an 
+error. 
+
+If the script is complicated enough that it contains a usage / help display, then the following 
+pattern can be used:
+
+``` powershell
+[CmdletBinding(PositionalBinding=$false)]
+param (
+    [switch]$build = $false, 
+    [string]$msbuildDir = "",
+    [parameter(ValueFromRemainingArguments=$true)] $badArgs)
+
+try {
+    if ($badArgs -ne $null) {
+        Print-Usage 
+        exit 1
+    }
+}
+
+```
+
+## Execution
+
+Invoking windows commands should be done via the Exec function.  This adds automatic error detection 
```

Comment:
You should probably add a note that this only works for executables that guarantee a return value of not-zero is a failure. This wrapping will not work for something like `robocopy` (just an example, `robocopy` should never be called from powershell anyways), which returns values other than 0 to represent success.